### ISSUE TITLE
Consider reprojecting to WGS84 automatically

### ISSUE THREAD

**Issue by hadley** [rstudio/leaflet#349]
```R
library(leaflet)
library(sp)

data("meuse", package = "sp", envir = environment())
sp::coordinates(meuse) <- ~x+y
proj4string(meuse) <- CRS("+init=epsg:28992")
meuse <- spTransform(meuse, CRS("+proj=longlat +datum=WGS84"))

leaflet(meuse) %>% 
  addCircles()
```

**Comment by jcheng5** [comment]
Considered and rejected a while back https://github.com/rstudio/leaflet/issues/139

**Comment by bhaskarvk** [comment]
Automatic reprojecting is a big NO-NO from the GIS guys. Almost everyone I talked to about this said don't do it. I think the reason is that not all data can be projected to WGS84 losslessly. I will let  @tim-salabim @mdsumner add their thoughts.

**Comment by hadley** [comment]
I think leaflet should at least give an informative error message. Re-projecting might be bad, but I think it's worse to silently accept an incompatible CRS.

**Comment by bhaskarvk** [comment]
That is a fair point. For all the add* methods, we can a) ensure that the shapefile's CRS is epsg4326 OR in case of just lat/lon values ensure that they are within correct lat/lon bounds.

But a major issue at times is that shapefiles don't always have correct CRS defined. So it might say 4326 but the data might not really be in that CRS. That's what Robert is saying in #139. So even if we check the CRS we are still not really sure whether the data is correct or not. That onus still lies on the user.

**Comment by hadley** [comment]
Yeah, so we'll warn and if they don't like the warning message, they'll have to correct the metadata.

**Comment by bhaskarvk** [comment]
One thing to keep in mind is, that for L.CRS.Simple CRS the input values for vector data can be anything, and don't necessarily have to be in lat/long. For all other projections including the default the input values for vector data have to be in lat/long. So the check needs to explicitly exclude L.CRS.Simple.

**Comment by bhaskarvk** [comment]
Also my second point was that the shapefile's CRS may say EPSG4326 (lat/long) in which case we won't warn but the actual data might not be in lat/lon and so it will not render correctly. No way to fix that.

**Comment by hadley** [comment]
Re-open to remind that we need warnings.  I think we should also include an example like above to illustrate what people should do

**Comment by mdsumner** [comment]
@bhaskarvk thanks for the ping here, I've tried to be measured in what I write below ;) 

We absolutely should just reproject when needed.  This is the most important thing that is missing across all of R spatial!  GIS generally doesn't have reproducible workflows, so you can easily get in trouble for the rare cases when 1) reprojection is applied to the source data  2) the actual transformation is not reversible.  The problem of incorrect metadata is a universal issue we face because so much data is sent around with not unit or crs metadata, it's not a reason to hobble modern tools. 

The best thing about GIS[1] is having a single map that you can throw everything into and it just puts it in the right place. This is one of the best features of `mapview` too, you give it metadata-aware objects and it puts it into its metadata-aware viewport.  The coordinate system used is a property of the map, and a usage requirement is that data coming in has the right metadata. This is used extensively in modern systems like GDAL and Manifold GIS, operations don't choke on data in different coordinate systems, they reproject to a common one and finish the task. 

1)  I have to caveat this, there's no such thing as "GIS" as if it's a single thing, it's a hot-bed of dozens of competing softwares and ideas.  R currently has the best overall transcendent capability, but it's fragmented and hardly uniform. 

The raster package works like this (for the most part, it will warn when reprojecting when doing extractions), and ggplot2 obviously works like this. It's the big hole in R graphics that the viewports don't have a standard metadata registration for the coordinate system in use.  Some packages do, I believe mapproj and oce do have a global registration for the current plot, it's something that R is sorely missing. 

If our data structures cannot store long-lat and X-Y side by side that's a different kind of problem, but really there should be no issue with reprojecting when needed. Simple features now can store multiple geometries, and for table-forms of data it's trivial to record the original coordinate values in extra columns (that's something that would be valuable for the actual times that reprojection is lossy). 

Personally, I see visualization and extractions and geometric/topological modifications as operations that extend and augment a data set - it's the business of the user to decide how to preserve the original data, and there are many options available to them.  Putting in limitations to protect users from themselves based on conservative old ideas will just hold us all back. Failing fast is critical, and the best way to see failure is to visualize it. 

The alternative, and I'd argue *we don't see this often* because it's so painful to do and get right is workflows like this (: 

```R
## PSEUDOCODE


prj <- "+proj=laea +long_0=147 ..." whatever my favourite projection
a <- readshapefile("a.shp")
b <- readshapefile("b.shp")

## hmm which one needs reprojecting?
## no error, no warning
plot(a); plot(b, add = TRUE)

## easy, just use the common one, but awkward to have to transform everything
plot(spTransform(a, prj)); plot(spTransform(b, prj), add = TRUE)

## here we may need to transform to a third common map
plot(somemap); plot(a, add = TRUE); plot(b, add = TRUE)

```

Mapview makes this all much better. 

```R
## PSEUDOCODE

## common, the right place to change the map is in the first mapview()
mapview(projection = prj) %>% mapview(a) %>% mapview(b)

## or at the end as a global viewport transformation (currently not well supported but logically sound)
mapview() %>% mapview(a) %>% mapview(b) %>% coordsys(projection = prj)
```

**Comment by tim-salabim** [comment]
Another aspect of this just came up in an [sf issue](https://github.com/edzer/sfr/issues/363). This is regarding the fact that `proj4string`s cannot be checked literally, only sematically as

`"+proj=longlat +ellps=WGS84 +no_defs"` 
is equivalent (sematically valid) to
`"+proj=longlat +datum=WGS84 +no_defs"`

So, [checking only for `"+datum=WGS84"`](https://github.com/rstudio/leaflet/blob/master/R/normalize-sf.R#L86) is not enough in some cases.
I think leaflet should either check for all possibly valid `longlat` expressions or not emit a warning at all. See [here](https://github.com/r-spatial/mapview/commit/e97c7af5dda55784c0a37f62c2de94f44b2e3db0#diff-5798b2525d12114a21aa684a5bb163f8R120) for a possible way to check for sematically valid `lonlat` definitions.
